0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Changing the behavior of an existing concept in a backwards-incompatible way always carries a high cost,
+and I do not propose it lightly. It optimizes for a better new Dafny user experience,
+where simple code that works with strings will be more correct by default,
+at the expense of existing users potentially needing to change their code when upgrading,
+or adding the `/unicodeChar:0` option to their build process.
+It is likely that the `/unicodeChar:0` option will need to be supported for a long time,
+to support users unable or unwilling to do this work,
+although the burden of maintaining this option should not be high.
+
+I propose this approach since Dafny is still relatively young as a programming language,
+and hence the impact of the breaking change is still relatively small.
+If anything, the longer we wait to correct this issue,
+the more Dafny code will continue to be written with subtle bugs that verification does not detect,
+or even new compilation targets that introduce them when compiling.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+There are many ways to tackle this problem, many of them inspired in part by how other programming languages
+have tackled them in the past (see also [Prior Art](#prior-art)).
+
+## Enforce valid UTF-16 strings in the verifier
+
+The cheapest solution to implement would likely be to add a new requirement in the verifier
+that surrogate `char` values are only used in pairs as intended.
+This would likely require more user effort to migrate their code, however,
+as every operation on strings would now have to prove that a surrogate pair
+is not being split up. 
+It also does not reduce the complexity of compiling
+strings to various target languages that do not necessarily use UTF-16.
+
+## Define "valid UTF-16 strings" as a subset type in a shared library
+
+The next cheapest solution would be to define a subset type of `string` that enforces
+correct usage of surrogate characters.
+Similar such definitions already exist for valid [UTF-8](https://github.com/aws/aws-encryption-sdk-dafny/blob/mainline/src/Util/UTF8.dfy#L21)
+and [UTF-16](https://github.com/dafny-lang/libraries/blob/master/src/Unicode/Utf16EncodingForm.dfy) code unit sequences
+in some Dafny codebases.
+This also introduces the same proof obligations as the previous solution,
+although here they could be proved by numerous helper methods that operate on the subset type
+in the same shared library.
+
+This would likely be the best solution if changing the Dafny language itself was not an option.
+The major downside is that built-in sequence operations such as `s[i]` and `s + s'` would still require additional
+effort to achieve verification, or would have to be abandoned entirely in favour of the helper methods
+provided by the shared library.
+
+## Change the definition of the `string` type
```

Comment:
The section describes downsides of having a standalone `string` type, but what are the upsides? Would it be possible to represent a dafny `abstractstring` using a native string datatype, while keeping it performant, if for example it had a limited API that excludes index based character access?